python/ray/tune/ray_trial_executor.py

```diff
@@ -216,7 +217,13 @@ def get_running_trials(self):
         return list(self._running.values())
 
     def get_next_available_trial(self):
-        [result_id], _ = ray.wait(list(self._running))
+        shuffled_results = random.sample(
+            self._running.keys(), len(self._running))
+        # Note: We shuffle the results because `ray.wait` by default returns
+        # the first available result, and we want to guarantee that slower
+        # trials (i.e. trials that run remotely) also get fairly reported.
+        # See https://github.com/ray-project/ray/issues/4211 for details.
+        [result_id], _ = ray.wait(shuffled_results)
```

Comment:
@ericl I don't think this needs a timeout. If I understand correctly, it's not the timeout for `ray.wait()` that matters in https://github.com/ray-project/ray/issues/1128 but the `time.sleep()` on the worker side. It seems like timeout wouldn't really make difference here. Correct me if I'm wrong.

Reply:
You're probably right. The timeout was only relevant due to the backend bug. Since that is now fixed, the cause of the wait unfairness is probably just the ordering